typescript/lib/parsers/openai.ts

```diff
@@ -700,17 +720,32 @@ export class OpenAIChatModelParser extends ParameterizedModelParser<Chat.ChatCom
     const output = aiConfig.getLatestOutput(prompt);
     if (output != null) {
       if (output.output_type === "execute_result") {
-        // If the prompt has output saved, add it to the messages array
         if (output.metadata?.rawResponse?.role === "assistant") {
-          if (typeof output.data === "string") {
-            messages.push({
-              content: output.data,
-              // TODO (rossdanlm): Support function_call and don't rely on rawResponse
-              role: output.metadata?.rawResponse?.role,
-              function_call: output.metadata?.rawResponse?.function_call,
-              name: output.metadata?.rawResponse?.name,
-            });
+          output as ExecuteResult;
+          let content: string | null = null;
+          let function_call: FunctionData | undefined = undefined;
+
+          if (output.data?.hasOwnProperty("value")) {
+            const outputData = output.data as OutputDataWithValue;
+            if (typeof outputData.value === "string") {
+              content = outputData.value;
+            } else if (outputData.kind === "tool_calls") {
+              outputData as OutputDataWithToolCallsValue;
+              // Typescript schema does not support array of function calls yet,
+              // but Python does so doing this for forward compatibility
+              function_call =
+                outputData.value[outputData.value.length - 1].function;
+            }
+          } else if (typeof output.data === "string") {
+            content = output.data;
           }
+
+          messages.push({
+            content,
+            role: output.metadata?.rawResponse?.role,
+            ...(function_call !== undefined && { function_call }),
+            name: output.metadata?.rawResponse?.name,
```

Comment:
What is `name`? And similar to `role`, can we do
`output.metadata?.name ?? output.metadata?.rawResponse?.name`?

Reply:
I'm not really seeing `name` in the OpenAI api

Reply:
It's defined in all of the sub-types for `ChatCompletionMessageParam` except for the tool one: https://github.com/openai/openai-node/blob/master/src/resources/chat/completions.ts#L445

But yes that's a good point I should check first whether it's undefined or not before setting, thanks for flagging